topics/w4-structure-editors.md

```diff
@@ -1,10 +1,40 @@
 
-#### Optional Reading
-* [Polymorphic Blocks](https://cseweb.ucsd.edu/\~lerner/poly-blocks.html): Lerner, S., Foster, S. R., & Griswold, W. G. (2015, April). Polymorphic blocks: Formalism-inspired UI for structured connectors. In Proceedings of the 33rd Annual ACM Conference on Human Factors in Computing Systems (pp. 3063-3072).
-* [Envision](https://www.pm.inf.ethz.ch/research/envision.html): Asenov, D., Hilliges, O., & Müller, P. (2016, May). The effect of richer visualizations on code comprehension. In Proceedings of the 2016 CHI Conference on Human Factors in Computing Systems (pp. 5040-5045).
+Early works in the field of structure editors include the Cornell Program Synthesizer[^2], in which the authors note that programs are not simply text, but instead complex syntactical structures, which should also be edited, executed, and debugged as such. Perhaps the most well-known structure editor is Scratch[^3], which allows users to drag-and-drop program blocks to form a program. Researchers have tried a number of variations on this concept, including hybrid approaches such as Envision[^4], which allows users to switch between graphical and textual representations of code, and quickly scaffold common program structures with keyboard commands. Unfortunately, such visual code editors get impractical and slow when working with larger projects in professional environments. The workflow they provide is simply too rigid for efficient use. For this reason, programmers try to minimize the use of graphical tools, and work exclusively through textual interfaces (looking at you, Vim users... 👀).
```

Comment:
It's a good coverage of work on the history of structure editing before the hazel paper